tracer/src/Datadog.Tracer.Native/fault_tolerant_rewriter.cpp

```diff
@@ -0,0 +1,594 @@
+    Logger::Info("Successfully added injection of successful instrumentation report, moduleId = ", moduleId);
+    return hr;
+}
+
+HRESULT FaultTolerantRewriter::RewriteInternal(RejitHandlerModule* moduleHandler,
+                                               RejitHandlerModuleMethod* methodHandler,
+                                               ICorProfilerFunctionControl* pFunctionControl)
+{
+    if (!is_fault_tolerant_instrumentation_enabled)
+    {
+        return m_methodRewriter->Rewrite(moduleHandler, methodHandler, pFunctionControl);
+    }
+
+    const auto moduleId = moduleHandler->GetModuleId();
+    const auto methodId = methodHandler->GetMethodDef();
+
+    if (FaultTolerantTracker::Instance()->IsKickoffMethod(moduleId, methodId))
+    {
+        const auto instrumentationVersion = m_methodRewriter->GetInstrumentationVersion(moduleHandler, methodHandler);
+        const auto instrumentingProduct = m_methodRewriter->GetInstrumentingProduct(moduleHandler, methodHandler);
+
+        if (FaultTolerantTracker::Instance()->IsInstrumentationVersionSucceeded(moduleId, methodId, instrumentationVersion, instrumentingProduct))
+        {
+            // Request Revert for the original method
+            //std::vector<MethodIdentifier> requests = {{moduleId, methodId}};
+            //auto promise = std::make_shared<std::promise<void>>();
+            //auto future = promise->get_future();
+            //m_rejit_handler->EnqueueRequestRevert(requests, promise);
+            //future.get();
+
+            // We also set the original IL function body, in case the revert did not succeeded.
+            // We do that just as a safeguard, it has harmless effect.
+            //const auto [pMethodBytes, methodSize] = FaultTolerantTracker::Instance()->GetILBodyAndSize(moduleId, methodId);
+            //pFunctionControl->SetILFunctionBody(methodSize, pMethodBytes);
+
+            return m_methodRewriter->Rewrite(moduleHandler, methodHandler, pFunctionControl);
+        }
+        else
+        {
+            return ApplyKickoffInstrumentation(moduleHandler, methodHandler, pFunctionControl);   
+        }
+    }
+    else if (FaultTolerantTracker::Instance()->IsOriginalMethod(moduleId, methodId))
+    {
+        return ApplyOriginalInstrumentation(moduleHandler, methodHandler, pFunctionControl);
+    }
+    else if (FaultTolerantTracker::Instance()->IsInstrumentedMethod(moduleId, methodId))
+    {
+        const auto methodIdOfKickoff =
+            FaultTolerantTracker::Instance()->GetKickoffMethodFromInstrumentedMethod(moduleId, methodId);
+        const auto [pMethodBytes, methodSize] =
+            FaultTolerantTracker::Instance()->GetILBodyAndSize(moduleId, methodIdOfKickoff);
+
+        pFunctionControl->SetILFunctionBody(methodSize, pMethodBytes);
+
+        // substitute the methodHandler of the instrumented duplication with the original (the one of the kickoff)
+        if (!moduleHandler->TryGetMethod(methodIdOfKickoff, &methodHandler))
```

Comment:
I guess that getting the original methodHandler makes the rest of the instrumentations oblivious to the fact that this is happening. Can you predict any ways that this could cause failures in the other rewriters (e.g. `TracerMethodRewriter` and `DebuggerMethodRewriter`)?

Reply:
I actually don't see why it will be an issue: the `DebuggerMethodRewriter` is already connected to the Fault-Tolerant and it works, I would guess the same for `TracerMethodRewriter`.
If it will come up as an issue I guess I will have to proxy that `moduleHandler` in one way or another before passing it in.